implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java

```diff
@@ -118,14 +131,21 @@ public boolean rawValueEquals(String name, String expected) {
      */
     public String getRawValue(String name) {
         for (ConfigSource configSource : getConfigSources()) {
-            String value = configSource.getValue(name);
+            String value = applyInterceptors(ConfigValue.of(configSource, name, configSource.getValue(name)));
             if (value != null) {
                 return value;
             }
         }
         return null;
     }
 
+    private String applyInterceptors(final ConfigValue configValue) {
+        final SmallRyeConfigSourceInterceptorContext context = new SmallRyeConfigSourceInterceptorContext(
```

Comment:
We don't want to construct a new context on every request for a configuration value; there should be a constant, pre-built context for every stage of the interception.  Relatedly, we want to apply the interceptors *before* acquiring the value from the config source.

So the invocation sequence should be something like:

```
-> getRawValue(name) (returns a ConfigValue now) calls `return ctxt0.proceed(name)`
   -> interceptor context 0 calls `return interceptor.getValue(ctxt1, name);`
      -> user interceptor 0 (can return a ConfigValue immediately or call proceed to...)
         -> interceptor context 1 calls `return interceptor.getValue(ctxt2, name);`
            -> user interceptor 1 ...
               -> interceptor context 2 (the last one, which calls...) ...
                  -> our "final" interceptor, which always consists of:
                      for (ConfigSource configSource : getConfigSoruces()) {
                          String value = configSource.getValue(name);
                          if (value != null) return new ConfigValue(name, value, ...);
                      }
                      return null;
```

Reply:
I think we are considering two different solutions.

My idea was to intercept after the getRawValue (so having the original value from the ConfigSource) and then allow the user to either proceed with logging, validation, etc. or provide a new value, by performing substitution or looking up other configs.

If I understood correctly, your proposal is to do it with the key instead?

Reply:
I'm considering the following requirements:

* Minimal API
* Allow for non-allocating or minimally-allocating implementations
* Avoid requiring thread locals or other "hidden" context passing
* Maximize flexibility

To avoid allocating, the interceptor chain has to be static (fixed).  But to have a static interceptor chain, any variable context (i.e. the name) has to be passed in as a thread-local value or a parameter.  Thread local values have implementation overhead in addition to being potentially bug-prone.  So, that essentially leaves what I've proposed above.

Given the above, I didn't notice any benefit to differentiating between the "original" configuration key and "additional" keys.  Furthermore I think there's a potential benefit to allowing interceptors to be able to "short circuit" the configuration value lookup process, returning a value (or lack of value) before any config source work is done.

If an interceptor wants the "original" value (from its perspective), it can call `proceed` with the original key name that was passed in, and acquire it that way.  At that point it can do whatever it wants with it - e.g. transform it, combine it with more values (via `proceed`).  Alternatively, if an interceptor doesn't care about the original value (or maybe it's better to say "the value associated with the original key"), it doesn't have to cause any work to be done to access it.

For example, we can implement Quarkus-style configuration profiles using an interceptor like this:

```java
public class ProfileInterceptor implements ConfigSourceInterceptor {
    private final String profileName = ....;
    public ConfigValue getValue(ConfigSourceInterceptorContext ctxt, String name) {
        ConfigValue v = ctxt.proceed("%" + profileName + "." + name);
        if (v == null) v = ctxt.proceed(name);
        return v;
    }
}
```

If a value is found for the profile-specific key, then the config source does not even need to be queried for the "original" value.  Furthermore, if another interceptor transforms the name some other way, who is to say what the "original" value even is?

Reply:
I always thought about the key being something immutable and that only the value should be allowed to be transformed. I like the Profile sample, on the other hand, I'm not sure about allowing to change the key. There is already a lot of confusion around key loading and key sources and developers trying to understand where the key came from. If for some reason you change the key name dramatically you might not even able to properly map your code, with configs and logs. You need to have very good logging.

So for me, keeping the original values and the transformed values, would at least give me some clear indication that a transformation happen, but I still now the origin and where I can change it.

Anyway, let me prototype it and see how it looks like.